src/natcap/invest/coastal_blue_carbon/coastal_blue_carbon.py

```diff
@@ -188,941 +194,1521 @@
+        Two ``dict``s, each of which maps string keys "soil" and "disturbance"
+        to ``scipy.sparse.dok_matrix`` objects of numpy.float32.  The first
+        such dict contains disturbance magnitudes for the pool for each
+        landcover transition, and the second contains accumulation rates for
+        the pool for the landcover transition.
     """
-    template_raster = C_prior_raster
-
-    T_s_rasters = []
-    A_r_rasters = []
-    E_r_rasters = []
-    N_r_rasters = []
-
-    for snapshot_idx in range(len(snapshot_years)):
-        snapshot_year = snapshot_years[snapshot_idx]
-        T_s_rasters.append(_OUTPUT['carbon_stock'] % (snapshot_year))
-        if snapshot_idx < len(snapshot_years)-1:
-            next_snapshot_year = snapshot_years[snapshot_idx + 1]
-            A_r_rasters.append(_OUTPUT['carbon_accumulation'] % (
-                snapshot_year, next_snapshot_year))
-            E_r_rasters.append(_OUTPUT['cabon_emissions'] % (
-                snapshot_year, next_snapshot_year))
-            N_r_rasters.append(_OUTPUT['carbon_net_sequestration'] % (
-                snapshot_year, next_snapshot_year))
-
-    # Total Net Sequestration
-    N_total_raster = 'total_net_carbon_sequestration.tif'
-
-    raster_registry_dict = {
-        'T_s_rasters': T_s_rasters,
-        'A_r_rasters': A_r_rasters,
-        'E_r_rasters': E_r_rasters,
-        'N_r_rasters': N_r_rasters,
-        'N_total_raster': N_total_raster,
+    table = utils.read_csv_to_dataframe(transition_csv_path, index_col=False)
+
+    lulc_class_to_lucode = {}
+    max_lucode = 0
+    for (lucode, values) in biophysical_dict.items():
+        lulc_class_to_lucode[values['lulc-class']] = lucode
+        max_lucode = max(max_lucode, lucode)
+
+    # Load up a sparse matrix with the transitions to save on memory usage.
+    # The number of possible rows/cols is the value of the maximum possible
+    # lucode we're indexing with plus 1 (to account for 1-based counting).
+    n_rows = max_lucode + 1
+    soil_disturbance_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    biomass_disturbance_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    soil_accumulation_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+    biomass_accumulation_matrix = scipy.sparse.dok_matrix(
+        (n_rows, n_rows), dtype=numpy.float32)
+
+    # I don't actually know if this uses less memory than the dict-based
+    # approach we had before since that, too, was basically sparse.
+    # At least this approach allows numpy-style indexing for easier 2D
+    # reclassifications.
+    # If we really wanted to save memory, we wouldn't duplicate the float32
+    # values here and instead use the transitions to index into the various
+    # biophysical values when reclassifying. That way we rely on python's
+    # assumption that ints<2000 or so are singletons and thus use less memory.
```

Comment:
I'm also missing the memory issue here. I'd normally expect that a user would have orders of magnitude fewer landcover classes than pixels in the raster. Is the concern that there'd be a unique lulc per pixel?

Reply:
Yes!  In fact, I ran across this very issue once, and it was due to my own (foolish) suggestion about how to work around the model's prior inability to handle spatial variation in biophysical parameters.  That's now addressed with having a second entrypoint for 'advanced' usage.